torch/csrc/distributed/spmd/event_handler_impl.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <c10d/ProcessGroup.hpp>
+#include <torch/csrc/distributed/spmd/event_handler.h>
+#include <torch/csrc/distributed/spmd/event_impl.h>
+
+
+namespace torch {
+namespace distributed {
+namespace spmd {
+
+// WIP DefaultTrigger
+class TORCH_API DefaultTrigger : public EventHandler {
+ public:
+  using EventHandler::EventHandler;
+  C10_NODISCARD std::vector<EventSchema> ingressEvents() const override;
+  C10_NODISCARD std::vector<EventSchema> egressEvents() const override;
+  std::vector<std::shared_ptr<Future>> handleEvent(
+      const c10::intrusive_ptr<Event>& event) override;
+
+ private:
+  std::vector<std::shared_ptr<Future>> handlePrepareModule(
+      c10::intrusive_ptr<PrepareModuleEvent> event);
+
+  std::vector<std::shared_ptr<Future>> handlePreForward(
+      c10::intrusive_ptr<PreForwardEvent> event);
+
+  void autogradHook(size_t index);
+
+  // keep grad accumulators alive
+  std::vector<std::shared_ptr<torch::autograd::Node>> gradAccumulators_;
+  std::vector<at::Tensor> params_;
+  std::vector<std::shared_ptr<Future>> gradReadyFutures_;
+};
+
+// WIP DefaultBucketer
+// TODO:
+// 1. split this into BucketIndexer and BucketAllocator
+// 2. bucketing gradients into larger tensors
```

Comment:
I'm assuming we will address most of the WIPs in future PRs to keep this one small?

Reply:
yep, that's the current plan. Let me know if any of these absolutely needs to be added to this PR, I can certainly add.